### ISSUE TITLE
Plugin Update/Uninstall is failing on Windows

### ISSUE THREAD

**Issue by grravity** [jellyfin/jellyfin#1623]
**Describe the bug**

'Check For Plugin Updates Failed'

**To Reproduce**

1. Go to 'Scheduled Tasks'
2. See error

**Log**

```
Access to the path 'Jellyfin.Plugin.Anime.dll' is denied.
at System.IO.FileSystem.RemoveDirectoryRecursive(String fullPath, WIN32_FIND_DATA& findData, Boolean topLevel)
at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
at Emby.Server.Implementations.Updates.InstallationManager.PerformPackageInstallation(IProgress`1 progress, String target, PackageVersionInfo package, CancellationToken cancellationToken)
at Emby.Server.Implementations.Updates.InstallationManager.InstallPackageInternal(PackageVersionInfo package, Boolean isPlugin, IProgress`1 progress, CancellationToken cancellationToken)
at Emby.Server.Implementations.Updates.InstallationManager.InstallPackage(PackageVersionInfo package, Boolean isPlugin, IProgress`1 progress, CancellationToken cancellationToken)
at Emby.Server.Implementations.ScheduledTasks.PluginUpdateTask.Execute(CancellationToken cancellationToken, IProgress`1 progress)
at Emby.Server.Implementations.ScheduledTasks.ScheduledTaskWorker.ExecuteInternal(TaskOptions options)
```

**Expected behavior**

Expected to properly update lol.

**Screenshots**

https://b.catgirlsare.sexy/tboM.png

**System:**
 - OS: [Windows 10]
 - Browser: [Firefox]
 - Jellyfin Version: [10.3.7]

**Comment by Z3ROGRAV1TYx** [comment]
Just now realizing it didn't update to 10.3.7, I wonder if I wnet about the update process wrong..
Does anyone have proper instructions for updating Jellyfin on Windows 10?

**Comment by cvium** [comment]
Looks like a permission issue.

**Comment by Z3ROGRAV1TYx** [comment]
Yeah, but it shouldn't have permission issues. I installed it under admin.

**Comment by anthonylavado** [comment]
What happens if you remove the anime plugin and install it again?

**Comment by Z3ROGRAV1TYx** [comment]
So, I tried with OpenSubtitles plugin first.

I uninstalled it, upon clicking uninstall it just showed a buffering icon and just sat there. I gave it a bit, then refreshed the page. It was gone. I went to catalog to reinstall it and clicked install and it just buffered again. Gave it time and eventually refreshed the page, went to check if it was installed and it didn't. I tried it a few more times no cigar.

I then tried installing another plugin (One that was never installed before), and it installed in seconds no buffer symbol. 

I tried going to the catalog and installing the latest anime plugin (Clicking the latest release then clicking install), without removing the old version (I assume it would just update it) but it just buffers and buffers. 

So I'm not sure what's going on.. I really don't want to do a Jellyfin reinstall though .-.

**Comment by anthonylavado** [comment]
Sounds like you’ll need to actually delete the plugins yourself from the folder they are located in. After that, you can try installing them again. You don’t have to completely reinstall Jellyfin.

**Comment by Z3ROGRAV1TYx** [comment]
@anthonylavado This will remove all preferences though correct? It will probabl work, I'll try later, but This shouldnt need to be done to simply update a plugin without error. Seems like something else is going on.

**Comment by anthonylavado** [comment]
The preferences should have been in separate files, I had meant for you only to remove the .DLL files.

At any rate, something changed from when it was first installed to now, otherwise there wouldn’t have been a permission error. When doing an update, the program actually goes and downloads the new plugin, deletes the old one, and puts the new one in its place. The error message “Access to the path 'Jellyfin.Plugin.Anime.dll' is denied” definitely sums up what was happening. If it can’t do it, it can’t do it.

I do apologize, but it’s one of the few things that’s out of our control. The best we can do is log the error and not crash when it happens :-(

**Comment by anthonylavado** [comment]
Reading back a little more - I’m not sure about what context Jellyfin is running in (this depends on how you installed it/started it up), but it’s entirely possible that you set it up as Admin, but it’s running as you/a different user now. Hence why it couldn’t delete what was originally set up as Admin. Just thinking out loud.

**Comment by Z3ROGRAV1TYx** [comment]
>The preferences should have been in separate files, I had meant for you only to remove the .DLL files.

Yes I saw they were in a seperate folder, I copied them in case, and then deleted them, reinstalled. Then they were still in that folder, they didn't save. I then tried replacing them with the ones I backed up, didn't work either for some reason.



>It’s entirely possible that you set it up as Admin, but it’s running as you/a different user now. Hence why it couldn’t delete what was originally set up as Admin. Just thinking out loud.

Unless it changed something by itself, it was installed as Admin, and there isn't even any other accounts on this computer.

**Comment by dkanada** [comment]
What user and group is jellyfin running as when you open htop or a similar program? What are the user, group, and file permissions for OpenSubtitles.dll and the plugin directory?

**Comment by anthonylavado** [comment]
Use Task Manager since you’re on Windows

**Comment by Z3ROGRAV1TYx** [comment]
> What user and group is Jellyfin running as when you open htop or a similar program? What are the user, group, and file permissions for OpenSubtitles.dll and the plugin directory?

> Use Task Manager since you’re on Windows

It's running under 'SYSTEM' if that's what you're asking.
It has admin permissions, I only have one account on this computer and it's an admin account. But yeah in the group it shows admins, my account and System.

**Comment by oddstr13** [comment]
This issue has also been reported [by another user on reddit][1].

### Steps to reproduce
 - Install plugin
 - Restart server
 - Attempt to uninstall plugin

(credit @anthonylavado)

The result is an infinite loading screen, and an error in the logfile.

### Workaround
 - Stop the `Jellyfin Server` service
 - Delete the offending plugin from `C:\ProgramData\Jellyfin\Server\plugins\`
 - Start the `Jellyfin Server` service again

[1]: https://old.reddit.com/r/jellyfin/comments/dcw4ma/things_that_need_to_be_improved_on_fixed/f2dlslr/

-----

**It does not seem to be a filesystem permission issue.**
The service is running as `NETWORK SERVICE`, and that is also the owner of `C:\ProgramData\Jellyfin\` with all it's content.
Also has list and traverse permissions on `C:\` and `C:\ProgramData\`.

```
[2019-10-05 00:09:57.692 +02:00] [INF] Deleting plugin directory "C:\ProgramData\Jellyfin\Server\plugins\Bookshelf"
[2019-10-05 00:09:57.805 +02:00] [ERR] Error processing request
System.UnauthorizedAccessException: Access to the path 'Jellyfin.Plugin.Bookshelf.dll' is denied.
   at System.IO.FileSystem.RemoveDirectoryRecursive(String fullPath, WIN32_FIND_DATA& findData, Boolean topLevel)
   at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
   at Emby.Server.Implementations.Updates.InstallationManager.UninstallPlugin(IPlugin plugin)
   at MediaBrowser.Api.PluginService.Delete(UninstallPlugin request)
   at Emby.Server.Implementations.Services.ServiceExecGeneral.<>c__DisplayClass5_0.<CreateExecFn>b__0(Object service, Object request)
   at Emby.Server.Implementations.Services.ServiceExecGeneral.Execute(Type serviceType, IRequest request, Object instance, Object requestDto, String requestName)
   at Emby.Server.Implementations.Services.ServiceController.Execute(HttpListenerHost httpHost, Object requestDto, IRequest req)
   at Emby.Server.Implementations.Services.ServiceHandler.ProcessRequestAsync(HttpListenerHost httpHost, IRequest httpReq, HttpResponse httpRes, ILogger logger, CancellationToken cancellationToken)
   at Emby.Server.Implementations.HttpServer.HttpListenerHost.RequestHandler(IHttpRequest httpReq, String urlString, String host, String localPath, CancellationToken cancellationToken)
```

**Comment by anthonylavado** [comment]
So the reason the plugin file can't be deleted is because it's in use. I'll give you one guess as to what process is tying up the file...

**Comment by anthonylavado** [comment]
So as far as I can tell, this has never worked in Windows. I just tested both 10.4(nightly-2019-10-04), 10.3.7, and 10.2.0 (where we first introduced plugins).

**Comment by anthonylavado** [comment]
Not sure where else to go here. @Bond-009 or @cvium, any ideas?

**Comment by n00bhax** [comment]
I can confirm that this problem has always been there. I have to completely delete a plugin manually to update one.

**Comment by JustAMan** [comment]
This is because how Windows itself works (it doesn't allow you to overwrite a file which is an executable or a DLL when the application using this exe/DLL still runs).

We don't have a proper fix for that yet...

**Comment by mark-monteiro** [comment]
Did a quick bit of research on this and loading/unloading DLLs at runtime is now supported in .NET Core 3.0. The plugin DLLs will need to be loaded via [AssemblyLoadContext](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-3.1). Then when updating, the assembly can be unloaded again with `AssemblyLoadContext.Unload`.

The assembly is unloaded asynchronously whenever garbage collection completes so it might be difficult to delete the old assembly file right away. Maybe a better option would be to create a new subfolder for each plugin version and then delete the old one with a scheduled task or next server startup, etc.

**Comment by mark-monteiro** [comment]
FYI I took a stab at this to see if we could use `AssemblyLoadContext` approach, but there is a bug with the XML serialization assembly that prevents this from working that isn't going to be fixed until .NET 5: https://github.com/dotnet/runtime/issues/1388. So it looks like well have to take another approach.

My branch is here in case anyone wants to take a look: https://github.com/jellyfin/jellyfin/compare/master...mark-monteiro:test-assembly-load-context

**Comment by BaronGreenback** [comment]
Don't know if this is of any use, but according to what i've read the MoveFileEx windows API call will allow dll's that are in use to be renamed. 

???
using System.Runtime.InteropServices;
[DllImport("Kernel32.dll", CharSet = CharSet32.Unicode)]
public static extern bool MoveFileEx(string lpExistingFileName, string lpNewFileName, int dwFlags);
[CODE]
MoveFileEx(file, file + ".bak", 4); // MOVEFILE_DELAY_UNTIL_REBOOT

Haven't looked at the code in detail to see how this could be implimented (if at all)

**Comment by JustAMan** [comment]
The problem with `MoveFileEx()` is we want to move the file away immediately, not upon next computer reboot, and also that flag requires administrative privileges which we cannot take for granted. Thus it doesn't seem to be a solution, alas :(

**Comment by BaronGreenback** [comment]
The admin privilege is a problem - but you can rename the file immediately with MoveFileEx() - just not delete it immediately.

"There are several applications, such as service packs and hotfixes, that must replace a file that's in use and is unable to. Windows therefore provides the MoveFileEx API to rename or delete a file and allows the caller to specify that they want the operation to take place the next time the system boots, before the files are referenced." - from sysinternals docs.

**Comment by JustAMan** [comment]
Again, we need to update the files without rebooting the server. I understand that in Windows world reboot is a universal solution, but rebooting a server just because one application in it wanted to update a plugin is still a bit of an overkill... I think we'd need to revise how we update plugins and streamline that instead.

**Comment by BaronGreenback** [comment]
Sorry - not making myself clear.

The first call to MoveFileEx would be as a rename and would permit you to copy a new version into the plugins folder. The old file will still be present and locked - but won't technically be in use as the code unloaded it.
The second call to MoveFileEx will mark it for deletion at the next reboot.

Combined with a random number on the rename, it could theoretically be done indefinitely - with all the old versions hanging around until the next reboot - at which point they would all be deleted.

eg.
movefileex - renames locked subs.dll -> subs1919192.bak (file remains locked)
new version copied into the folder - and dynamcally loaded.
movefileex - marks the locked file subs1919192.bak for deletion at the next reboot
rinse and repeat.

**Comment by BaronGreenback** [comment]
Design change that would solve the problem.  #3401

**Comment by smidley** [comment]
I can confirm that I've had this issue with every version of JF server.  https://github.com/jellyfin/jellyfin/issues/4052

**Comment by mark-monteiro** [comment]
@BaronGreenback Was this resolved with #3401?

**Comment by BaronGreenback** [comment]
yes
